core/lib/src/rocket.rs

```diff
@@ -641,135 +781,79 @@ impl Rocket {
-    }
-
-    // TODO.async: Solidify the Listener APIs and make this function public
-    async fn listen_on<L>(mut self, listener: L) -> Result<(), crate::error::Error>
-    where
-        L: Listener + Send + Unpin + 'static,
-        <L as Listener>::Connection: Send + Unpin + 'static,
-    {
-        self = self.prelaunch_check().map_err(crate::error::Error::Launch)?;
-
-        self.fairings.pretty_print_counts();
-
-        // Determine the address and port we actually binded to.
-        self.config.port = listener.local_addr().map(|a| a.port()).unwrap_or(0);
-
-        let proto = if self.config.tls.is_some() {
-            "https://"
-        } else {
-            "http://"
-        };
-
-        let full_addr = format!("{}:{}", self.config.address, self.config.port);
-
-        // Set the keep-alive.
-        // TODO.async: implement keep-alive in Listener
-        // let timeout = self.config.keep_alive.map(|s| Duration::from_secs(s as u64));
-        // listener.set_keepalive(timeout);
-
-        // Freeze managed state for synchronization-free accesses later.
-        self.state.freeze();
-
-        // Run the launch fairings.
-        self.fairings.handle_launch(&self);
-
-        launch_info!("{}{} {}{}",
-                     Paint::masked("🚀 "),
-                     Paint::default("Rocket has launched from").bold(),
-                     Paint::default(proto).bold().underline(),
-                     Paint::default(&full_addr).bold().underline());
-
-        // Restore the log level back to what it originally was.
-        logger::pop_max_level();
-
-        // We need to get this before moving `self` into an `Arc`.
-        let mut shutdown_receiver = self.shutdown_receiver
-            .take().expect("shutdown receiver has already been used");
-
-        let rocket = Arc::new(self);
-        let service = hyper::make_service_fn(move |connection: &<L as Listener>::Connection| {
-            let rocket = rocket.clone();
-            let remote_addr = connection.remote_addr().unwrap_or_else(|| ([0, 0, 0, 0], 0).into());
-            async move {
-                Ok::<_, std::convert::Infallible>(hyper::service_fn(move |req| {
-                    hyper_service_fn(rocket.clone(), remote_addr, req)
-                }))
-            }
-        });
-
-        #[derive(Clone)]
-        struct TokioExecutor;
-
-        impl<Fut> hyper::Executor<Fut> for TokioExecutor where Fut: Future + Send + 'static, Fut::Output: Send {
-            fn execute(&self, fut: Fut) {
-                tokio::spawn(fut);
+    // NB: Not an `async fn` because _attach() calls finish(), recursing.
+    pub(crate) fn finish(&mut self) -> BoxFuture<'_, ()> {
+        Box::pin(async move {
+            while !self.pending.is_empty() {
+                let op = self.pending.remove(0);
+                let manifest = self.manifest.take().expect("internal error: manifest was taken and not replaced. Was `inspect()` called but not polled to completion?");
```

Comment:
This needs to be broken into multiple lines. Use `\` to break up a string without adding space:

```rust
"this is a very \
long string"
```